infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,223 @@
+        this.streamLog = streamLog;
+        this.addressSpaceView = addressSpaceView;
+    }
+
+    @Override
+    public CompletableFuture<BatchResult> transfer(Batch batch) {
+        return readRecords(batch, 0)
+                .thenApply(records -> writeRecords(records, streamLog))
+                .exceptionally(error -> BatchResult
+                        .builder()
+                        .status(FAILED)
+                        .destinationServer(batch.getDestination())
+                        .addresses(batch.getAddresses())
+                        .build()
+                );
+    }
+
+    /**
+     * Reads data entries by utilizing the replication protocol. If there are errors, retries.
+     *
+     * @param batch The batch of addresses to be read.
+     * @param retries The number of retries.
+     * @return A result of reading records.
+     */
+    CompletableFuture<ReadBatch> readRecords(Batch batch, int retries) {
+        return CompletableFuture
+                .supplyAsync(() -> {
+                            Supplier<Map<Long, ILogData>> readSupplier = () -> addressSpaceView
+                                    .simpleProtocolRead(batch.getAddresses(), readOptions);
+
+                            return Result
+                                    .of(readSupplier)
+                                    .mapError(BatchProcessorError::new);
+                        }
+                )
+                .thenApply(readResult ->
+                        readResult.map(records -> checkReadRecords(
+                                batch.getAddresses(),
+                                records,
+                                batch.getDestination()
+                        ))
+                )
+                .thenCompose(checkedReadResult -> {
+                    if (checkedReadResult.isError()) {
+                        return retryReadRecords(batch, retries);
+                    } else {
+                        ReadBatch readBatch = checkedReadResult.get();
+                        if (readBatch.getStatus() == ReadBatch.FailureStatus.FAILED) {
+                            return retryReadRecords(readBatch.toBatch(), retries);
+                        } else {
+                            return CompletableFuture.completedFuture(readBatch);
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Handle read errors. Exponential backoff policy is utilized.
+     *
+     * @param batch        Batch of addresses to be read.
+     * @param retriesTried Configurable number of retries for the current batch.
+     * @return Future of the result of the retry.
+     */
+    CompletableFuture<ReadBatch> retryReadRecords
+    (Batch batch, int retriesTried) {
+        try {
+            AtomicInteger readRetries = new AtomicInteger(retriesTried);
+            return IRetry.build(ExponentialBackoffRetry.class, RetryExhaustedException.class, () -> {
```

Comment:
IRetry - looks awkward, let's try to provide a better implementation for retriable logic

Reply:
IRetry will be redone in a separate fix